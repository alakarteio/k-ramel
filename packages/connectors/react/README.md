# @k-ramel/react
> React bindings for k-ramel

 - Main repository: [k-ramel](https://github.com/alakarteio/k-ramel)

## Examples & API
### provide the store to react context
```jsx
import { provider } from '@k-ramel/react'
import store from './store' // see k-ramel documentation about building your store

const App = () => <div>My application</div>

export default provider(store)(App)
```

### map store to your components
**component.container.js**
```jsx
import { inject } from '@k-ramel/react'
import Component from './component' // your react component

/* in this example we expect that the store is build with:
  createStore({
    data: {
      users: types.keyValue(),
    },
  })
*/
// this function allow you to access:
// - the whole store
// - the props given by the parent component
// - the drivers (see driver documentation if you don't know what it is)
const mapStore = (store, ownProps, drivers) => {
  // you should return an object
  // this object represent the new props to be merged with `ownProps` from parent
  // note that props written here will override parent ones
  return {
    // data
    ...store.data.users.get(ownProps.id),
    // callback that use a driver
    // - callback are not tested upon `shouldComponentUpdate` optimisation
    onRemove: () => {
      drivers.http('USERS').delete(`/api/users/${ownProps.id}`)
    },
    // The example above would be better this way
    onRemoveBetter: () => store.dispatch({Â type: '@@ui/USERS>REMOVED>CLICKED', payload: ownProps.id })
    // - then plug a reation to call the http driver
    // - this way your UI and your data management are decoupled
  },
}

export default inject(mapStore)(Component)
```

### write some logical code binded to a component
To understand this part you should first read the [listeners' documentation](../../k-ramel/doc/LISTENERS.md).

First, write your reactions:\
**component.reactions.js**
```js
export const callDeleteUser = (action, store, drivers) => {
  drivers.http('USERS').delete(`/api/users/${action.payload}`)
}

export const removeUser = (action, store) => {
  // here we will remove user from store AFTER the API respond
  // but you can do it in parallel (by writing this line into `callDeleteUser` for example)
  store.users.remove(action.payload.id)
}
```

Second, write your listeners:\
**component.listeners.js**
```js
// import listener engine
import { when } from 'k-ramel'
// import your reactions (logical code)
import { callDeleteuser, removeUserÂ } from './component.reactions'


// react connector needs an array of listeners
export default [
  // here you can see our UI and your logical code are decupled
  // the `@@ui/USERS>REMOVED>CLICKED` is dispatched from your component (mapStore)
  when('@@ui/USERS>REMOVED>CLICKED')(callDeleteUser),
  // when the http driver has a response, it triggers a `ENDED` event
  // you can see the full http driver to see the list of events
  // note that data is already parsed from JSON
  when('@@http/USERS>DELETE>ENDED')(removeUser),
]
```

Finally, you can attach your listeners to a given component.\
One of the strategy is to attach listeners to a `Screen` of your application.\
That means that your business code is attach to a particular screen:\
***component.container.js**
```js
// this is the previous component.container we call but we add some lines on it
// to plug ours listeners (and reactions)
// - first modification is here, we import `listen` to plug our listeners
import { inject, listen /* HERE ðŸ‘‹ */ } from '@k-ramel/react'
import Component from './component'
// we also import our listeners ðŸ‘‹
import listeners from './component.listeners'

const mapStore = (store, ownProps, drivers) => {
  return {
    ...store.data.users.get(ownProps.id),
    onRemoveBetter: () => store.dispatch({Â type: '@@ui/USERS>REMOVED>CLICKED', payload: ownProps.id })
  },
}

// we use `listen` as an HoC ðŸ‘‹
export default listen(
  listeners,  // [REQUIRED] we pass our listeners writen above ðŸ‘‹
  'users', // [OPTIONAL] you can pass a name to your listeners
)(inject(mapStore)(Component))
// from here, each time your component will be **mounted**:
// - an action will be dispatched: `@@krml/LISTENERS>ADDED>users` (`users` is the name we give just above)
// - the listeners are attached to the k-ramel store
// each time your component will be **unmountedÂ£Â£:
// - an action will be dispatched: `@@krml/LISTENERS>REMOVING>users` (`users` is the name we give just above)
// - the listeners will be removed from k-ramel store
//
// note that you can listen to `@@krml/LISTENERS>ADDED>users` in your plugued listeners
// since this action is dispatched AFTER listeners are added
// it means you can some screen initialisation for example
//
// also note that you can listen to `@@krml/LISTENERS>REMOVING>users` but you **CAN'T** trust it to reset your store
// because, by React nature, `ADDED` action from an other screen would be dispatched BEFORE the `REMOVING` one.
```
