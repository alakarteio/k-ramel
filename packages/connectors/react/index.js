!function(e,t){"object"==typeof exports&&"undefined"!=typeof module?t(exports,require("react"),require("fbjs/lib/shallowEqual")):"function"==typeof define&&define.amd?define(["exports","react","fbjs/lib/shallowEqual"],t):t(e["@k-ramel/react"]={},e.React,e["fbjs/lib/shallowEqual"])}(this,function(e,t,n){"use strict";var r="default"in t?t.default:t;n=n&&n.hasOwnProperty("default")?n.default:n;var o=function(e){return e.displayName||e.name||e.constructor&&e.constructor.name||"Unknown"},i=function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")},u=function(){function e(e,t){for(var n=0;n<t.length;n++){var r=t[n];r.enumerable=r.enumerable||!1,r.configurable=!0,"value"in r&&(r.writable=!0),Object.defineProperty(e,r.key,r)}}return function(t,n,r){return n&&e(t.prototype,n),r&&e(t,r),t}}(),c=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e},s=function(e,t){if("function"!=typeof t&&null!==t)throw new TypeError("Super expression must either be null or a function, not "+typeof t);e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,enumerable:!1,writable:!0,configurable:!0}}),t&&(Object.setPrototypeOf?Object.setPrototypeOf(e,t):e.__proto__=t)},l=function(e,t){if(!e)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return!t||"object"!=typeof t&&"function"!=typeof t?e:t},a={},f=function(e){return Object.keys(e).reduce(function(t,n){var r=e[n];return"function"==typeof r?t:c({},t,function(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}({},n,r))},a)},p=[];e.inject=function(e){return function(a){var p,h;return h=p=function(t){function o(t,n){i(this,o);var r=l(this,(o.__proto__||Object.getPrototypeOf(o)).call(this,t,n));return r.inject=function(t){r.setState(function(n){return c({},n,{injectedProps:e(r.context.store,t||r.props)})})},r.first=!0,r.state={injectedProps:{}},r}return s(o,t),u(o,[{key:"componentWillMount",value:function(){var e=this;this.unsubscribe=this.context.store.subscribe(function(){e.inject()}),this.inject()}},{key:"componentWillReceiveProps",value:function(e){this.inject(e)}},{key:"shouldComponentUpdate",value:function(e,t){return!!this.first||!(n(this.props,e)&&n(f(t.injectedProps),f(this.state.injectedProps)))}},{key:"componentWillUnmount",value:function(){this.unsubscribe()}},{key:"render",value:function(){return this.first&&(this.first=!1),r.createElement(a,c({},this.props,this.state.injectedProps))}}]),o}(t.Component),p.displayName="inject("+o(a),p.contextTypes={store:function(){return null}},h}},e.provider=function(e){return function(n){var c,a;return a=c=function(t){function o(){return i(this,o),l(this,(o.__proto__||Object.getPrototypeOf(o)).apply(this,arguments))}return s(o,t),u(o,[{key:"getChildContext",value:function(){return{store:e}}},{key:"componentWillMount",value:function(){e.dispatch("@@krml/INIT")}},{key:"render",value:function(){return r.createElement(n,this.props)}}]),o}(t.Component),c.displayName="provider("+o(n)+")",c.childContextTypes={store:function(){return null}},a}},e.listen=function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:p;return function(n){var c,a;return a=c=function(t){function o(){return i(this,o),l(this,(o.__proto__||Object.getPrototypeOf(o)).apply(this,arguments))}return s(o,t),u(o,[{key:"componentWillMount",value:function(){this.context.store.listeners.add(e)}},{key:"componentWillUnmount",value:function(){this.context.store.listeners.remove(e)}},{key:"render",value:function(){return r.createElement(n,this.props)}}]),o}(t.Component),c.displayName="listen("+o(n),c.contextTypes={store:function(){return null}},a}},Object.defineProperty(e,"__esModule",{value:!0})});
//# sourceMappingURL=index.js.map
