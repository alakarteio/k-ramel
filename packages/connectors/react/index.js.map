<<<<<<< HEAD
{"version":3,"file":"index.js","sources":["src/getWrappedDisplayName.js","src/inject.jsx","src/listen.jsx","src/provider.jsx"],"sourcesContent":["export default Component => (\n  Component.displayName\n  || Component.name\n  || (Component.constructor && Component.constructor.name)\n  || 'Unknown'\n)\n","import React, { Component } from 'react'\nimport shallowEqual from 'fbjs/lib/shallowEqual'\nimport getWrappedDisplayName from './getWrappedDisplayName'\n\nconst defaultObject = {}\nconst withoutFunctions = object => Object\n  .keys(object)\n  .reduce(\n    (acc, key) => {\n      const value = object[key]\n      if (typeof value === 'function') return acc\n      return { ...acc, [key]: value }\n    },\n    defaultObject,\n  )\n\nexport default injectFunction => WrappedComponent => class extends Component {\n  static displayName = `inject(${getWrappedDisplayName(WrappedComponent)}`\n\n  static contextTypes = {\n    store: () => null, // this is to avoid importing prop-types\n  }\n\n  constructor(props, context) {\n    super(props, context)\n\n    this.first = true\n    this.state = {\n      injectedProps: {},\n    }\n  }\n\n  componentWillMount() {\n    const { store } = this.context\n\n    if (!store) {\n      const bold = 'font-weight: bolder; font-style: italic;'\n      // eslint-disable-next-line no-console\n      console.error(\n        `[k-ramel/react] Error in %cinject%c for the component %c${getWrappedDisplayName(WrappedComponent)}%c\\n` +\n        '\\t> The store needs to be provided by an ancestor of this component.\\n' +\n        '\\t> You can use %cprovider%c from %c@k-ramel/react%c or %cProvider%c from %creact-redux%c.\\n\\n' +\n        'Check the documentation for an example at https://github.com/alakarteio/k-ramel#connect-it-with-reactjs\\n',\n        bold, '', bold, '', bold, '', bold, '', bold, '', bold, '',\n      )\n      return\n    }\n\n    // subscribe\n    this.unsubscribe = store.subscribe(() => {\n      this.inject()\n    })\n\n    // run in once\n    this.inject()\n  }\n\n  componentWillReceiveProps(nextProps) {\n    this.inject(nextProps)\n  }\n\n  shouldComponentUpdate(nextProps, nextState) {\n    if (this.first) return true\n\n    return !(\n      shallowEqual(this.props, nextProps)\n      && shallowEqual(\n        withoutFunctions(nextState.injectedProps),\n        withoutFunctions(this.state.injectedProps),\n      )\n    )\n  }\n\n  componentWillUnmount() {\n    this.unsubscribe()\n  }\n\n  inject = (nextProps) => {\n    this.setState(state => ({\n      ...state,\n      injectedProps: injectFunction(this.context.store, nextProps || this.props),\n    }))\n  }\n\n  render() {\n    if (this.first) this.first = false\n\n    return (\n      <WrappedComponent\n        /* this is parent props */\n        {...this.props}\n        /* this is injected props from hoc */\n        {...this.state.injectedProps}\n      />\n    )\n  }\n}\n","import React, { Component } from 'react'\nimport getWrappedDisplayName from './getWrappedDisplayName'\n\nconst defaultListeners = []\n\nconst toActionFactory = (name) => {\n  const suffix = `${name ? '>' : ''}${name || ''}`\n  return type => `@@krm/LISTENERS>${type}${suffix}`\n}\n\nexport default (listeners = defaultListeners, name) => (WrappedComponent) => {\n  const toAction = toActionFactory(name)\n\n  return class extends Component {\n    static displayName = `listen(${getWrappedDisplayName(WrappedComponent)}`\n\n    static contextTypes = {\n      store: () => null, // this is to avoid importing prop-types\n    }\n\n    componentWillMount() {\n      const { store } = this.context\n\n      store.listeners.add(listeners)\n      store.dispatch(toAction('ADDED'))\n    }\n\n    componentWillUnmount() {\n      const { store } = this.context\n\n      store.dispatch(toAction('REMOVING'))\n      store.listeners.remove(listeners)\n    }\n\n    render() {\n      return <WrappedComponent {...this.props} />\n    }\n  }\n}\n","/* eslint-disable import/prefer-default-export */\nimport React, { Component } from 'react'\nimport getWrappedDisplayName from './getWrappedDisplayName'\n\nexport default store => WrappedComponent => class extends Component {\n  static displayName = `provider(${getWrappedDisplayName(WrappedComponent)})`\n\n  static childContextTypes = {\n    store: () => null, // this is to avoid importing prop-types\n  }\n\n  getChildContext() {\n    return { store }\n  }\n\n  componentWillMount() {\n    store.dispatch('@@krml/INIT')\n  }\n\n  render() {\n    return <WrappedComponent {...this.props} />\n  }\n}\n"],"names":["Component","displayName","name","constructor","defaultObject","withoutFunctions","Object","keys","object","reduce","acc","key","value","defaultListeners","props","context","inject","nextProps","setState","state","injectFunction","_this","store","first","this","unsubscribe","subscribe","bold","error","getWrappedDisplayName","WrappedComponent","nextState","shallowEqual","injectedProps","React","contextTypes","dispatch","childContextTypes","listeners","toAction","suffix","type","toActionFactory","add","remove"],"mappings":"0YAAA,yBACEA,EAAUC,aACPD,EAAUE,MACTF,EAAUG,aAAeH,EAAUG,YAAYD,MAChD,m+BCACE,KACAC,EAAmB,mBAAUC,OAChCC,KAAKC,GACLC,OACC,SAACC,EAAKC,OACEC,EAAQJ,EAAOG,SACA,mBAAVC,EAA6BF,OAC5BA,4HAAMC,EAAMC,KAE1BR,ICVES,iCDa2B,sDAOnBC,EAAOC,4EACXD,EAAOC,aAqDfC,OAAS,SAACC,KACHC,SAAS,wBACTC,iBACYC,EAAeC,EAAKN,QAAQO,MAAOL,GAAaI,EAAKP,cAtDjES,OAAQ,IACRJ,oGAMGG,EAAUE,KAAKT,QAAfO,SAEHA,OAcAG,YAAcH,EAAMI,UAAU,aAC5BV,gBAIFA,kBAlBGW,EAAO,mDAELC,MACN,2DAA2DC,EAAsBC,uRAIjFH,EAAM,GAAIA,EAAM,GAAIA,EAAM,GAAIA,EAAM,GAAIA,EAAM,GAAIA,EAAM,uDAcpCV,QACnBD,OAAOC,iDAGQA,EAAWc,WAC3BP,KAAKD,SAGPS,EAAaR,KAAKV,MAAOG,IACtBe,EACD3B,EAAiB0B,EAAUE,eAC3B5B,EAAiBmB,KAAKL,MAAMc,qEAM3BR,sDAWDD,KAAKD,QAAOC,KAAKD,OAAQ,GAG3BW,gBAACJ,OAEKN,KAAKV,MAELU,KAAKL,MAAMc,uBA5E4CjC,eAC1DC,sBAAwB4B,EAAsBC,KAE9CK,oBACE,kBAAM,yCEhBO,iNAQXb,wDAIHc,SAAS,uDAIRF,gBAACJ,EAAqBN,KAAKV,cAhBoBd,eACjDC,wBAA0B4B,EAAsBC,SAEhDO,yBACE,kBAAM,mCDEDC,yDAAYzB,EAAkBX,sBAAS,SAAC4B,WAChDS,EANgB,SAACrC,OACjBsC,GAAYtC,EAAO,IAAM,KAAKA,GAAQ,WACrC,qCAA2BuC,EAAOD,GAIxBE,CAAgBxC,gMAUrBoB,EAAUE,KAAKT,QAAfO,QAEFgB,UAAUK,IAAIL,KACdF,SAASG,EAAS,6DAIhBjB,EAAUE,KAAKT,QAAfO,QAEFc,SAASG,EAAS,eAClBD,UAAUM,OAAON,2CAIhBJ,gBAACJ,EAAqBN,KAAKV,cAtBjBd,eACZC,sBAAwB4B,EAAsBC,KAE9CK,oBACE,kBAAM"}
=======
{"version":3,"file":"index.js","sources":["src/getWrappedDisplayName.js","src/inject.jsx","src/listen.jsx","src/provider.jsx"],"sourcesContent":["export default Component => (\n  Component.displayName\n  || Component.name\n  || (Component.constructor && Component.constructor.name)\n  || 'Unknown'\n)\n","import React, { Component } from 'react'\nimport shallowEqual from 'fbjs/lib/shallowEqual'\nimport getWrappedDisplayName from './getWrappedDisplayName'\n\nconst defaultObject = {}\nconst withoutFunctions = object => Object\n  .keys(object)\n  .reduce(\n    (acc, key) => {\n      const value = object[key]\n      if (typeof value === 'function') return acc\n      return { ...acc, [key]: value }\n    },\n    defaultObject,\n  )\n\nexport default injectFunction => WrappedComponent => class extends Component {\n  static displayName = `inject(${getWrappedDisplayName(WrappedComponent)}`\n\n  static contextTypes = {\n    store: () => null, // this is to avoid importing prop-types\n  }\n\n  constructor(props, context) {\n    super(props, context)\n\n    this.first = true\n    this.state = {\n      injectedProps: {},\n    }\n  }\n\n  componentWillMount() {\n    // subscribe\n    this.unsubscribe = this.context.store.subscribe(() => {\n      this.inject()\n    })\n\n    // run in once\n    this.inject()\n  }\n\n  componentWillReceiveProps(nextProps) {\n    this.inject(nextProps)\n  }\n\n  shouldComponentUpdate(nextProps, nextState) {\n    if (this.first) return true\n\n    return !(\n      shallowEqual(this.props, nextProps)\n      && shallowEqual(\n        withoutFunctions(nextState.injectedProps),\n        withoutFunctions(this.state.injectedProps),\n      )\n    )\n  }\n\n  componentWillUnmount() {\n    this.unsubscribe()\n  }\n\n  inject = (nextProps) => {\n    this.setState(state => ({\n      ...state,\n      injectedProps: injectFunction(this.context.store, nextProps || this.props),\n    }))\n  }\n\n  render() {\n    if (this.first) this.first = false\n\n    return (\n      <WrappedComponent\n        /* this is parent props */\n        {...this.props}\n        /* this is injected props from hoc */\n        {...this.state.injectedProps}\n      />\n    )\n  }\n}\n","import React, { Component } from 'react'\nimport getWrappedDisplayName from './getWrappedDisplayName'\n\nconst defaultListeners = []\n\nconst toActionFactory = (name) => {\n  const suffix = `${name ? '>' : ''}${name || ''}`\n  return type => `@@krml/LISTENERS>${type}${suffix}`\n}\n\nexport default (listeners = defaultListeners, name) => (WrappedComponent) => {\n  const toAction = toActionFactory(name)\n\n  return class extends Component {\n    static displayName = `listen(${getWrappedDisplayName(WrappedComponent)}`\n\n    static contextTypes = {\n      store: () => null, // this is to avoid importing prop-types\n    }\n\n    componentWillMount() {\n      const { store } = this.context\n\n      store.listeners.add(listeners)\n      store.dispatch(toAction('ADDED'))\n    }\n\n    componentWillUnmount() {\n      const { store } = this.context\n\n      store.dispatch(toAction('REMOVING'))\n      store.listeners.remove(listeners)\n    }\n\n    render() {\n      return <WrappedComponent {...this.props} />\n    }\n  }\n}\n","/* eslint-disable import/prefer-default-export */\nimport React, { Component } from 'react'\nimport getWrappedDisplayName from './getWrappedDisplayName'\n\nexport default store => WrappedComponent => class extends Component {\n  static displayName = `provider(${getWrappedDisplayName(WrappedComponent)})`\n\n  static childContextTypes = {\n    store: () => null, // this is to avoid importing prop-types\n  }\n\n  getChildContext() {\n    return { store }\n  }\n\n  componentWillMount() {\n    store.dispatch('@@krml/INIT')\n  }\n\n  render() {\n    return <WrappedComponent {...this.props} />\n  }\n}\n"],"names":["Component","displayName","name","constructor","defaultObject","withoutFunctions","Object","keys","object","reduce","acc","key","value","defaultListeners","props","context","inject","nextProps","setState","state","injectFunction","_this","store","first","unsubscribe","this","subscribe","nextState","shallowEqual","injectedProps","React","WrappedComponent","getWrappedDisplayName","contextTypes","dispatch","childContextTypes","listeners","toAction","suffix","type","toActionFactory","add","remove"],"mappings":"0YAAA,yBACEA,EAAUC,aACPD,EAAUE,MACTF,EAAUG,aAAeH,EAAUG,YAAYD,MAChD,m+BCACE,KACAC,EAAmB,mBAAUC,OAChCC,KAAKC,GACLC,OACC,SAACC,EAAKC,OACEC,EAAQJ,EAAOG,SACA,mBAAVC,EAA6BF,OAC5BA,4HAAMC,EAAMC,KAE1BR,ICVES,iCDa2B,sDAOnBC,EAAOC,4EACXD,EAAOC,aAsCfC,OAAS,SAACC,KACHC,SAAS,wBACTC,iBACYC,EAAeC,EAAKN,QAAQO,MAAOL,GAAaI,EAAKP,cAvCjES,OAAQ,IACRJ,yGAOAK,YAAcC,KAAKV,QAAQO,MAAMI,UAAU,aACzCV,gBAIFA,2DAGmBC,QACnBD,OAAOC,iDAGQA,EAAWU,WAC3BF,KAAKF,SAGPK,EAAaH,KAAKX,MAAOG,IACtBW,EACDvB,EAAiBsB,EAAUE,eAC3BxB,EAAiBoB,KAAKN,MAAMU,qEAM3BL,sDAWDC,KAAKF,QAAOE,KAAKF,OAAQ,GAG3BO,gBAACC,OAEKN,KAAKX,MAELW,KAAKN,MAAMU,uBA7D4C7B,eAC1DC,sBAAwB+B,EAAsBD,KAE9CE,oBACE,kBAAM,yCEhBO,iNAQXX,wDAIHY,SAAS,uDAIRJ,gBAACC,EAAqBN,KAAKX,cAhBoBd,eACjDC,wBAA0B+B,EAAsBD,SAEhDI,yBACE,kBAAM,mCDEDC,yDAAYvB,EAAkBX,sBAAS,SAAC6B,WAChDM,EANgB,SAACnC,OACjBoC,GAAYpC,EAAO,IAAM,KAAKA,GAAQ,WACrC,sCAA4BqC,EAAOD,GAIzBE,CAAgBtC,gMAUrBoB,EAAUG,KAAKV,QAAfO,QAEFc,UAAUK,IAAIL,KACdF,SAASG,EAAS,6DAIhBf,EAAUG,KAAKV,QAAfO,QAEFY,SAASG,EAAS,eAClBD,UAAUM,OAAON,2CAIhBN,gBAACC,EAAqBN,KAAKX,cAtBjBd,eACZC,sBAAwB+B,EAAsBD,KAE9CE,oBACE,kBAAM"}
>>>>>>> master
