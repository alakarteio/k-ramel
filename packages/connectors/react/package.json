{
  "name": "@k-ramel/react",
  "version": "3.0.0",
  "main": "dist/index.umd.js",
  "module": "dist/index.es.js",
  "author": "Fabien JUIF <fabien.juif@gmail.com>",
  "license": "MIT",
  "description": "React bindings for k-ramel",
  "repository": "https://github.com/alakarteio/k-ramel/tree/master/packages/connectors/react",
  "dependencies": {
    "fbjs": "^1.0.0"
  },
  "devDependencies": {
    "cross-env": "^7.0.0",
    "k-ramel": "^3.0.0"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "types": "./dist/index.d.ts"
}
