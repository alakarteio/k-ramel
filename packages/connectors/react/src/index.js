export { default as inject } from './inject'
export { default as provider } from './provider'
export { default as listen } from './listen'
