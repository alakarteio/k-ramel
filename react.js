!function(e,t){"object"==typeof exports&&"undefined"!=typeof module?t(exports,require("react"),require("fbjs/lib/shallowEqual")):"function"==typeof define&&define.amd?define(["exports","react","fbjs/lib/shallowEqual"],t):t(e["k-simple-state-react"]={},e.React,e.shallowEqual)}(this,function(e,t,n){"use strict";var r="default"in t?t.default:t;n=n&&n.hasOwnProperty("default")?n.default:n;var o=function(e){return e.displayName||e.name||e.constructor&&e.constructor.name||"Unknown"},i=function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")},u=function(){function e(e,t){for(var n=0;n<t.length;n++){var r=t[n];r.enumerable=r.enumerable||!1,r.configurable=!0,"value"in r&&(r.writable=!0),Object.defineProperty(e,r.key,r)}}return function(t,n,r){return n&&e(t.prototype,n),r&&e(t,r),t}}(),c=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e},s=function(e,t){if("function"!=typeof t&&null!==t)throw new TypeError("Super expression must either be null or a function, not "+typeof t);e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,enumerable:!1,writable:!0,configurable:!0}}),t&&(Object.setPrototypeOf?Object.setPrototypeOf(e,t):e.__proto__=t)},a=function(e,t){if(!e)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return!t||"object"!=typeof t&&"function"!=typeof t?e:t};e.inject=function(e){return function(l){var f,p;return p=f=function(t){function o(t,r){i(this,o);var u=a(this,(o.__proto__||Object.getPrototypeOf(o)).call(this,t,r));return u.inject=function(t){var r=u.state.injectedProps,o=e(u.context.store,t||u.props);n(r,o)||u.setState(function(e){return c({},e,{injectedProps:o})})},u.state={injectedProps:{}},u}return s(o,t),u(o,[{key:"componentWillMount",value:function(){var e=this;this.unsubscribe=this.context.store.subscribe(function(){e.inject()}),this.inject()}},{key:"componentWillReceiveProps",value:function(e){this.inject(e)}},{key:"componentWillUnmount",value:function(){this.unsubscribe()}},{key:"render",value:function(){return r.createElement(l,c({},this.props,this.state.injectedProps))}}]),o}(t.Component),f.displayName="inject("+o(l),f.contextTypes={store:function(){return null}},p}},e.provider=function(e){return function(n){var c,l;return l=c=function(t){function o(){return i(this,o),a(this,(o.__proto__||Object.getPrototypeOf(o)).apply(this,arguments))}return s(o,t),u(o,[{key:"getChildContext",value:function(){return{store:e}}},{key:"render",value:function(){return r.createElement(n,this.props)}}]),o}(t.Component),c.displayName="provider("+o(n)+")",c.childContextTypes={store:function(){return null}},l}},Object.defineProperty(e,"__esModule",{value:!0})});
//# sourceMappingURL=react.js.map
