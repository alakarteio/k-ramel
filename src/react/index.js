export { default as inject } from './inject'
export { default as provider } from './provider'
export { default as withListeners } from './withListeners'
