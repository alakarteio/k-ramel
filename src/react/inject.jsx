import React, { Component } from 'react'
import shallowEqual from 'fbjs/lib/shallowEqual'
import getWrappedDisplayName from './getWrappedDisplayName'

const withoutFunctions = object => Object
  .keys(object)
  .reduce((acc, key) => {
    const value = object[key]
    if (typeof value === 'function') return acc
    return { ...acc, [key]: value }
  })


export default injectFunction => WrappedComponent => class extends Component {
  static displayName = `inject(${getWrappedDisplayName(WrappedComponent)}`

  static contextTypes = {
    store: () => null, // this is to avoid importing prop-types
  }

  constructor(props, context) {
    super(props, context)

    this.state = {
      injectedProps: {},
    }
  }

  componentWillMount() {
    // subscribe
    this.unsubscribe = this.context.store.subscribe(() => {
      this.inject()
    })

    // run in once
    this.inject()
  }

  componentWillReceiveProps(nextProps) {
    this.inject(nextProps)
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  inject = (nextProps) => {
    const { injectedProps } = this.state
    const newInjectedProps = injectFunction(this.context.store, nextProps || this.props)

    if (
      shallowEqual(this.props, nextProps)
      && shallowEqual(
        withoutFunctions(injectedProps),
        withoutFunctions(newInjectedProps),
      )
    ) { return }

    this.setState(state => ({
      ...state,
      injectedProps: newInjectedProps,
    }))
  }

  render() {
    return (
      <WrappedComponent
        /* this is parent props */
        {...this.props}
        /* this is injected props from hoc */
        {...this.state.injectedProps}
      />
    )
  }
}
