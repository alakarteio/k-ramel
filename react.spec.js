/* eslint-env jest */
import * as lib from './react'
import tests from './misc/test-suite.react'

tests(lib)
