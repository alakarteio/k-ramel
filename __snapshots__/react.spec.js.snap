// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react should add listeners at mount 1`] = `
Object {
  "dispatch": Array [
    Array [
      "@@krm/LISTENERS>ADDING",
    ],
    Array [
      "@@krm/LISTENERS>ADDED",
    ],
    Array [
      "@@krm/LISTENERS>REMOVING",
    ],
    Array [
      "@@krm/LISTENERS>REMOVED",
    ],
  ],
}
`;

exports[`react should inject store to a component 1`] = `"<div>a label</div>"`;

exports[`react should provide store as context and dispatch an INIT event 1`] = `
Object {
  "dispatch": Array [
    Array [
      "@@krml/INIT",
    ],
  ],
  "html": "<div>{\\"this\\":\\"is my store\\"}</div>",
}
`;

exports[`react should refresh when props changes 1`] = `"<div>a label-2</div>"`;

exports[`react should refresh when store changes 1`] = `"<div>an other label</div>"`;
